### PR TITLE
makes HAMT iteration 10x faster

### PR DESCRIPTION
**Pull request by ruslansennov**
The HAMT iterator completely redesigned

**before**
```
Target            Operation   Impl                         Params  Count            Score  ±     Error    Unit
HashSetBenchmark  Iterate     slang_persistent               1000     18        4,299.731  ±     1.18%   ops/s
HashSetBenchmark  Iterate     slang_persistent                100     18       45,524.153  ±     1.77%   ops/s
HashSetBenchmark  Iterate     slang_persistent                 10     18      680,479.197  ±     1.31%   ops/s
HashSetBenchmark  Iterate     scala_persistent               1000     18      129,055.550  ±     0.24%   ops/s
HashSetBenchmark  Iterate     scala_persistent                100     18    1,437,694.285  ±     0.83%   ops/s
HashSetBenchmark  Iterate     scala_persistent                 10     18   22,035,129.813  ±     1.27%   ops/s
HashSetBenchmark  Iterate     pcollections_persistent        1000     18       44,336.905  ±     1.08%   ops/s
HashSetBenchmark  Iterate     pcollections_persistent         100     18      472,947.893  ±     1.33%   ops/s
HashSetBenchmark  Iterate     pcollections_persistent          10     18    3,469,540.882  ±     0.70%   ops/s

Performance Ratios
================================================================================================================
  (Outliers removed: 30.00% low end, 10.00% high end)

Ratios slang / <alternative_impl>
Target            Operation   Ratio                                                    10        100       1000 
HashSetBenchmark  Iterate     slang_persistent/pcollections_persistent              0.20x      0.10x      0.10x 
HashSetBenchmark  Iterate     slang_persistent/scala_persistent                     0.03x      0.03x      0.03x 

Ratios <alternative_impl> / slang
Target            Operation                                             Ratio          10        100       1000 
HashSetBenchmark  Iterate            pcollections_persistent/slang_persistent       5.10x     10.39x     10.31x 
HashSetBenchmark  Iterate                   scala_persistent/slang_persistent      32.38x     31.58x     30.01x 
```

**after**

```
Target            Operation   Impl                         Params  Count            Score  ±     Error    Unit
HashSetBenchmark  Iterate     slang_persistent               1000     18       49,484.074  ±     0.26%   ops/s
HashSetBenchmark  Iterate     slang_persistent                100     18      522,937.431  ±     0.98%   ops/s
HashSetBenchmark  Iterate     slang_persistent                 10     18    4,728,144.061  ±     1.99%   ops/s
HashSetBenchmark  Iterate     scala_persistent               1000     18      130,241.725  ±     0.38%   ops/s
HashSetBenchmark  Iterate     scala_persistent                100     18    1,595,782.555  ±     1.03%   ops/s
HashSetBenchmark  Iterate     scala_persistent                 10     18   22,418,589.165  ±     1.87%   ops/s
HashSetBenchmark  Iterate     pcollections_persistent        1000     18       45,362.049  ±     1.55%   ops/s
HashSetBenchmark  Iterate     pcollections_persistent         100     18      471,520.387  ±     0.25%   ops/s
HashSetBenchmark  Iterate     pcollections_persistent          10     18    3,433,121.007  ±     3.10%   ops/s

Performance Ratios
================================================================================================================
  (Outliers removed: 30.00% low end, 10.00% high end)

Ratios slang / <alternative_impl>
Target            Operation   Ratio                                                    10        100       1000 
HashSetBenchmark  Iterate     slang_persistent/pcollections_persistent              1.38x      1.11x      1.09x 
HashSetBenchmark  Iterate     slang_persistent/scala_persistent                     0.21x      0.33x      0.38x 

Ratios <alternative_impl> / slang
Target            Operation                                             Ratio          10        100       1000 
HashSetBenchmark  Iterate            pcollections_persistent/slang_persistent       0.73x      0.90x      0.92x 
HashSetBenchmark  Iterate                   scala_persistent/slang_persistent       4.74x      3.05x      2.63x 
```

### CODE REVIEW

File: javaslang/src/main/java/javaslang/collection/HashArrayMappedTrie.java [hunk 1]
```diff
@@ -51,6 +51,87 @@
         PUT, REMOVE
     }
 
+    class It<K, V> extends AbstractIterator<LeafNode<K, V>> {
```
Comment (danieldietrich) [inline javaslang/src/main/java/javaslang/collection/HashArrayMappedTrie.java]: Please name it `LeafNodeIterator`

File: javaslang/src/main/java/javaslang/collection/HashArrayMappedTrie.java [hunk 2]
```diff
@@ -101,6 +182,20 @@ static int fromBitmap(int bitmap, int bit) {
 
         abstract AbstractNode<K, V> modify(int shift, int keyHashCode, K key, V value, Action action);
 
+        Iterator<LeafNode<K, V>> nodes() {
+            return new It<>(this);
+        }
+
+        @Override
+        public Iterator<Tuple2<K, V>> iterator() {
+            return nodes().map(node -> Tuple.of(node.key(), node.value()));
```
Comment (danieldietrich) [inline javaslang/src/main/java/javaslang/collection/HashArrayMappedTrie.java]: Why is a node not backed by a Tuple?, i.e. s.th. like `nodes().map(AbstractNode::entry)`

File: javaslang/src/main/java/javaslang/collection/HashArrayMappedTrie.java [hunk 3]
```diff
@@ -51,6 +51,87 @@
...
+                Object[] subNodes = ((IndexedNode<K, V>) node).subNodes;
+                return index < subNodes.length ? (AbstractNode<K, V>) subNodes[index] : null;
+            } else if (node instanceof ArrayNode) {
+                ArrayNode<K, V> arrayNode = (ArrayNode<K, V>) node;
+                return index < 32 ? (AbstractNode<K, V>) arrayNode.subNodes[index] : null;
+            }
+            return null;
+        }
+    }
+
```
Comment (danieldietrich) [inline javaslang/src/main/java/javaslang/collection/HashArrayMappedTrie.java]: Wow, sophisticated Iterator :)

File: javaslang/src/main/java/javaslang/collection/HashArrayMappedTrie.java (hunk 2 above)
Comment (ruslansennov) [inline javaslang/src/main/java/javaslang/collection/HashArrayMappedTrie.java]: Morning
(please refresh page, PR is updated)

Well, this is good question. If node will be backed by Tuple, we will obtain better performance in iteration along tuples. On other side we'll spend more memory. I'll investigate this idea, but now I would leave this as is

File: javaslang/src/main/java/javaslang/collection/HashArrayMappedTrie.java (hunk 3 above)
Comment (ruslansennov) [inline javaslang/src/main/java/javaslang/collection/HashArrayMappedTrie.java]: Yep :smile: 
Last iterator was simple and obvious, but used a lot (O(n)) of temporary objects

File: javaslang/src/main/java/javaslang/collection/HashArrayMappedTrie.java (hunk 2 above)
Comment (danieldietrich) [inline javaslang/src/main/java/javaslang/collection/HashArrayMappedTrie.java]: Morning :)

Yes, if it makes sense it should be a separate PR

File: javaslang/src/main/java/javaslang/collection/HashArrayMappedTrie.java [hunk 4]
```diff
@@ -51,6 +51,88 @@
         PUT, REMOVE
     }
 
+    class LeafNodeIterator<K, V> extends AbstractIterator<LeafNode<K, V>> {
+
+        // buckets levels + leaf level = (Integer.SIZE / AbstractNode.SIZE + 1) + 1
+        private final static int MAX_LEVELS = Integer.SIZE / AbstractNode.SIZE + 2;
```
Comment (danieldietrich) [inline javaslang/src/main/java/javaslang/collection/HashArrayMappedTrie.java]: I don't understand the comment.

`(Integer.SIZE / AbstractNode.SIZE + 1) + 1` not equals `Integer.SIZE / AbstractNode.SIZE + 2`

Comment (ruslansennov) [inline javaslang/src/main/java/javaslang/collection/HashArrayMappedTrie.java]: really?! :trollface:

Comment (danieldietrich) [inline javaslang/src/main/java/javaslang/collection/HashArrayMappedTrie.java]: ok, then MAX_LEVELS is an upper bound, right?

Comment (danieldietrich) [inline javaslang/src/main/java/javaslang/collection/HashArrayMappedTrie.java]: why not set `MAX_LEVELS` to `(Integer.SIZE / AbstractNode.SIZE + 1) + 1`?

Comment (ruslansennov) [inline javaslang/src/main/java/javaslang/collection/HashArrayMappedTrie.java]: > ok, then MAX_LEVELS is an upper bound, right?

yes, just maximum value of tree depth

> why not set MAX_LEVELS to (Integer.SIZE / AbstractNode.SIZE + 1) + 1?

hmmm... I see no difference, but I can do it

Comment (danieldietrich) [inline javaslang/src/main/java/javaslang/collection/HashArrayMappedTrie.java]: no, leave it as is

**Comment by danieldietrich** [comment]
Thx!
